src/iop/ashift.c

```diff
@@ -5754,7 +5754,11 @@ void commit_params(struct dt_iop_module_t *self,
   d->orthocorr = (p->mode == ASHIFT_MODE_GENERIC) ? 0.0f : p->orthocorr;
   d->aspect = (p->mode == ASHIFT_MODE_GENERIC) ? 1.0f : p->aspect;
 
-  if(gui_has_focus(self))
+  if(gui_has_focus(self)
+     || isnan(p->cl)
+     || isnan(p->cr)
+     || isnan(p->ct)
+     || isnan(p->cb))
```

Comment:
Isn't @ralfbrown eliminating `isnan()` everywhere?

Reply:
Yes, but this is to recover from a bad XMP with NAN. The $MIN & $MAX also added to the introspection circuitry for those 4 fields should avoid new NAN but at least we can recover from broken XMP already created.

Reply:
isnan() might break after @ralfbrown introduces compiler flags for non-NAN maths

Reply:
It's a question for @ralfbrown I think

Reply:
> The $MIN & $MAX also added to the introspection circuitry for those 4 fields should avoid new NAN

No they won't. They are used to set the range for a slider, if one is connected to these param fields (which they aren't), and in the _optional_ call to `_iop_validate_params`, which needs to be enabled by `DT_DEBUG_PARAMS/-d params` but even then only does diagnostic checking and no _fixing_.

> isnan() might break after @ralfbrown introduces compiler flags for non-NAN maths

InputNG extensively uses `NAN/isnan()` for signalling purposes. I.e. if you _explicitly_ send a move size of `NAN ` if will only do a read and if an option is not available, a `NAN` is returned. So these NANs are _not_ the result of a calculation, just the defined value. Will that still break when the compiler flag is set?

Reply:
AFAIK yes

Reply:
Because `isnan` gets optimised away even when not looking at the result of a calc? Seems weird but ohkay.

Would it then be possible to have our own implementation that simply does `(int)x == (int)NAN` or something like that? Or is even NAN not defined anymore? I guess we could use G_MAXFLOAT or something if needed... Not sure how to test for that in lua.

Reply:
I plan to add a `dt_isnan(x) { return isnan(x); }` for those cases we can't eliminate.  It turns out that a function compiled with different optimization flags won't get inlined (even if declared "inline"), so it should be safe to set the finite-math-only flag globally if all the NAN checks are inside code compiled without it.

@dterrahe: I had noticed that lots of action callbacks check for NAN.  Since these checks are not done per pixel, an extra function call would be negligible.  But we could also switch the "read-only" signal to -FLT_MAX.

Reply:
> But we could also switch the "read-only" signal to -FLT_MAX.

Working on it. Only snag is that this also gets passed back to/through LUA, so LUA will need to have a way to check for -FLT_MAX. Haven't looked into that yet. A Q&D < 1e6 will work too, so not a show stopper. Checking for NAN in LUA goes like `value ~= value` :-)

Reply:
Should have @wpferguson; checking the return of `dt.gui.action` would need to change and requesting a read-only action by passing a NAN (0/0) as the last parameter will no longer be supported. However, since recently the last parameter can simply be dropped (or, in the rare case it is followed by a multiinstance number, set as "")